src/arizona_live_reload.erl

```diff
@@ -39,57 +37,69 @@ Live-reload functionality for use during development.
 -export([handle_info/2]).
 
 %% State
--record(state, {timer, files = #{}, clients = #{}}).
-
-%% Macros
--define(SERVER, ?MODULE).
+-record(state, {
+    timer :: undefined | reference(),
+    files :: #{string() := {erl, modified}},
+    clients :: #{pid() := boolean()}
+}).
+-opaque state() :: #state{}.
+-export_type([state/0]).
+-elvis([{elvis_style, state_record_and_type, disable}]). % opaque not identified as "type"
 
 %% --------------------------------------------------------------------
 %% API functions.
 %% --------------------------------------------------------------------
 
+-spec start_link() -> gen_server:start_ret().
 start_link() ->
-    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
 
+-spec reload() -> ok.
 reload() ->
-    gen_server:cast(?SERVER, reload).
+    gen_server:cast(?MODULE, reload).
 
 %% --------------------------------------------------------------------
 %% gen_server callbacks.
 %% --------------------------------------------------------------------
 
-init([]) ->
+-spec init(Args) -> {ok, State}
+    when Args :: term(),
+         State :: state().
+init(_Args) ->
     register_events(),
     setup_watcher(),
-    {ok, #state{}}.
-
-handle_call(Request, From, State) ->
-    ?LOG_INFO("[LiveReload] call: ~p from ~p~n", [Request, From]),
-    {reply, State, State}.
-
+    {ok, #state{
+        files = #{},
+        clients = #{}
+    }}.
+
+-spec handle_call(Request, From, State) -> no_return()
+    when Request :: term(),
+         From :: gen_server:from(),
+         State :: state().
+handle_call(_Request, _From, _State) ->
+    exit(not_implemented).
+
+-spec handle_cast(Request, State1) -> {noreply, State2}
+    when Request :: term(),
+         State1 :: state(),
+         State2 :: state().
 handle_cast(reload, State) ->
-    ?LOG_INFO("[LiveReload] reload: ~p~n", [State]),
     reload(State#state.clients),
-    {noreply, State};
-handle_cast(Request, State) ->
-    ?LOG_INFO("[LiveReload] cast: ~p~n", [Request]),
     {noreply, State}.
 
+-spec handle_info(Info, State1) -> {noreply, State2}
+    when Info :: term(),
+         State1 :: state(),
+         State2 :: state().
 handle_info({init, Client}, #state{clients = Clients} = State) ->
-    ?LOG_INFO("[LiveReload] init: ~p~n", [Client]),
     {noreply, State#state{clients = Clients#{Client => true}}};
 handle_info({terminate, Client}, #state{clients = Clients} = State) ->
-    ?LOG_INFO("[LiveReload] terminate: ~p~n", [Client]),
     {noreply, State#state{clients = maps:without([Client], Clients)}};
 handle_info({_Pid, {fs, file_event}, {File, Events}},
             #state{timer = Timer, files = Files} = State0) ->
     % Try recompile only when the last modified file was received.
-    case Timer =:= undefined of
-        true ->
-            ok;
-        false ->
-            erlang:cancel_timer(Timer)
-    end,
+    Timer =/= undefined andalso erlang:cancel_timer(Timer),
```

Comment:
I don't always remember that this is possible in Erlang :sweat_smile:

Reply:
I used to abuse it, in the past, and do stuff like

```erlang
<cond1> andalso
  begin
    <cond2> andalso
      begin
        <ret>
      end
  end
```

but then that meant you'd have to make sure all <cond> where boolean, and the spec would always be `false | <ret>`.

I started not liking this pattern so did https://github.com/inaka/elvis_core/pull/233.

In any case, `<cond> andalso <ret>` is not that bad when it's isolated.